### PR TITLE
kill OBJ_TAINT etc.

### PR DESCRIPTION
**Pull request by shyouhei**
Now that `RUBY_FL_TAINT` is recycled to become new `RUBY_FL_SHAREABLE`.
Setting/clearing this flag from extension libraries break Ractor. Especially problematic one is `OBJ_INFECT`, which would make non-shareable objects travel across Ractor boundaries.

Such operations should just be prohibited.

See also https://bugs.ruby-lang.org/issues/16131

### CODE REVIEW

File: include/ruby/internal/fl_type.h [hunk 1]
```diff
@@ -204,7 +209,7 @@ ruby_fl_type {
     RUBY_FL_SINGLETON = RUBY_FL_USER0,
 };
 
-enum { RUBY_FL_DUPPED = RUBY_T_MASK | RUBY_FL_EXIVAR | RUBY_FL_TAINT };
+enum { RUBY_FL_DUPPED = RUBY_T_MASK | RUBY_FL_EXIVAR | RUBY_FL_SHAREABLE };
```
Comment (shyouhei) [inline include/ruby/internal/fl_type.h:212]: @ko1 is this intentional that `Object#dup` inherits shareability? Or should we delete it?

**Review by jeremyevans — APPROVED** [review]
This looks good to me.  It does sound like the shareable flag should not be copied on `dup`, because duping a frozen object (sharable) should result in an unfrozen object (not sharable).  However, maybe this is handled another way?  Hopefully @ko1 knows.

File: include/ruby/internal/fl_type.h (hunk 1 above)
Comment (ko1) [reply include/ruby/internal/fl_type.h:230]: (1) You can delete RUBY_FL_SHAREABLE  here.
(2) RUBY_FL_DUPPED and rb_dup_setup which uses RUBY_FL_DUPPED are not used in core and maybe nobody uses them, so we can remove these definitions.


```
$ gem-codesearch rb_dup_setup
/srv/gems/rhodes-7.1.17/platform/shared/ruby/include/ruby/ruby.h:#define DUPSETU
/srv/gems/rhodes-7.1.17/platform/shared/ruby/include/ruby/ruby.h:rb_dup_setup(VA
/srv/gems/ruby-compiler-0.1.1/vendor/ruby/include/ruby/ruby.h:#define DUPSETUP(d
/srv/gems/ruby-compiler-0.1.1/vendor/ruby/include/ruby/ruby.h:rb_dup_setup(VALUE
/srv/gems/zscan-2.0.9/ext/pack/ruby/internal/newobj.h:#define DUPSETUP   rb_dup_
/srv/gems/zscan-2.0.9/ext/pack/ruby/internal/newobj.h:rb_dup_setup(VALUE dup, VA
```

Comment (shyouhei) [reply include/ruby/internal/fl_type.h:230]: OK, thanks for checking.

File: include/ruby/internal/fl_type.h [hunk 2]
```diff
@@ -165,9 +168,29 @@ ruby_fl_type {
     RUBY_FL_PROMOTED1    = (1<<6),
     RUBY_FL_PROMOTED     = RUBY_FL_PROMOTED0 | RUBY_FL_PROMOTED1,
     RUBY_FL_FINALIZE     = (1<<7),
-    RUBY_FL_TAINT        = (1<<8),
+    RUBY_FL_TAINT
+
+#if RBIMPL_HAS_EXTENSION(enumerator_attributes)
+    RBIMPL_ATTR_DEPRECATED(("taintedness turned out to be a wrong idea."))
+#elif RBIMPL_COMPILER_SINCE(GCC, 6, 0, 0)
+    RBIMPL_ATTR_DEPRECATED(("taintedness turned out to be a wrong idea."))
```
Comment (nobu) [inline include/ruby/internal/fl_type.h:176]: Why repeating the same message?

**Review by nobu — APPROVED** [review]

File: include/ruby/internal/fl_type.h (hunk 2 above)
Comment (shyouhei) [reply include/ruby/internal/fl_type.h:176]: Because GCC 6 doesn't understand `__has_extension()`, we have to do things once for Clang and once for GCC.

Comment (nobu) [reply include/ruby/internal/fl_type.h:176]: That means `RBIMPL_HAS_EXTENSION(enumerator_attributes) || RBIMPL_COMPILER_SINCE(GCC, 6, 0, 0)` doesn't work?

Comment (shyouhei) [reply include/ruby/internal/fl_type.h:176]: Yes, that should work.  That line would be a bit too long to me though.  It's just a matter of preference.  I like 80-ish columns.

File: include/ruby/internal/fl_type.h (hunk 1 above)
Comment (shyouhei) [reply include/ruby/internal/fl_type.h]: https://github.com/ruby/ruby/pull/4100/commits/9e083a24e1405bb23efe6c98aef95db03ef06a7d resolves this.

File: include/ruby/internal/fl_type.h [hunk 3]
```diff
@@ -204,7 +227,19 @@ ruby_fl_type {
...
-enum { RUBY_FL_DUPPED = RUBY_T_MASK | RUBY_FL_EXIVAR | RUBY_FL_TAINT };
+enum {
+    RUBY_FL_DUPPED
+
+#if RBIMPL_HAS_EXTENSION(enumerator_attributes)
+    RBIMPL_ATTR_DEPRECATED(("It seems there is no actual usage of this enum."))
+#elif RBIMPL_COMPILER_SINCE(GCC, 6, 0, 0)
+    RBIMPL_ATTR_DEPRECATED(("It seems there is no actual usage of this enum."))
+#elif defined(_MSC_VER)
+# pragma deprecated(RUBY_FL_UNTRUSTED)
```
Comment (nobu) [inline include/ruby/internal/fl_type.h:238]:
```suggestion
# pragma deprecated(RUBY_FL_DUPPED)
```

File: include/ruby/internal/fl_type.h (hunk 2 above)
Comment (shyouhei) [reply include/ruby/internal/fl_type.h]: https://github.com/ruby/ruby/pull/4100/commits/bb668e13e10c38dadcf8190078ef22be2c5d99df

Comment (nobu) [reply include/ruby/internal/fl_type.h]: I agree that it is too long.
OTOH, repeating same blocks three times also feels redundant to me.
I thought a macro for deprecated enums, such as `RBIMPL_DEPRECATED_ENUM(RUBY_FL_TAINT, ("taintedness turned out to be a wrong idea."))`, but MSC doesn't support `_Pragma`?

Comment (nobu) [reply include/ruby/internal/fl_type.h]: Seems nice.
Yet another question came out.
`RBIMPL_ATTR_DEPRECATED` seems always defined here, doesn't it work unconditionally?

Comment (shyouhei) [reply include/ruby/internal/fl_type.h]: No, some old GCC don't understand `enum { foo __attribute__((deprecated)) };`.

Comment (nobu) [reply include/ruby/internal/fl_type.h]: I see, thank you.

**Pull request merged**